File: src/devices/bin/driver_manager/composite/parent_set_collector.cc
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/devices/bin/driver_manager/composite/parent_set_collector.h"

#include <fidl/fuchsia.driver.framework/cpp/fidl.h>

#include "src/devices/bin/driver_manager/resource.h"
#include "src/devices/lib/log/log.h"

namespace driver_manager {

zx::result<> ParentSetCollector::AddNode(
    uint32_t index, const std::vector<fuchsia_driver_framework::NodeProperty2>& node_properties,
    ResourceWkPtr resource) {
  ZX_ASSERT(HasCompositeInfo());
  ZX_ASSERT(index < parents_.size());

  if (parents_[index] != std::nullopt && !parents_[index]->expired()) {
    return zx::error(ZX_ERR_ALREADY_BOUND);
  }
  parents_[index] = std::move(resource);
  parent_properties_[index] =
      fuchsia_driver_framework::NodePropertyEntry2(parent_names_.value()[index], node_properties);

  if (auto resource_ptr = parents_[index]->lock(); resource_ptr) {
    if (auto node_ptr = resource_ptr->owner().lock(); node_ptr) {
      node_ptr->MarkAsCompositeParent();
    }
  }

  return zx::ok();
}

void ParentSetCollector::ReleaseNodes() {
  for (auto& resource : parents_) {
    if (resource == std::nullopt) {
      continue;
    }
    if (auto resource_ptr = resource->lock(); resource_ptr) {
      if (auto node_ptr = resource_ptr->owner().lock(); node_ptr) {
        node_ptr->UnmarkAsCompositeParent();
      }
    }
    resource.reset();
  }
}

zx::result<std::shared_ptr<Node>> ParentSetCollector::TryToAssemble(
    std::string_view name, NodeManager* node_manager, async_dispatcher_t* dispatcher) {
  ZX_ASSERT(HasCompositeInfo());
  if (completed_composite_node_ && !completed_composite_node_->expired()) {
    return zx::error(ZX_ERR_ALREADY_EXISTS);
  }

  std::vector<NodeWkPtr> parent_nodes;
  parent_nodes.reserve(parents_.size());
  for (auto& resource : parents_) {
    if (resource == std::nullopt) {
      return zx::error(ZX_ERR_SHOULD_WAIT);
    }
    auto resource_ptr = resource.value().lock();
    if (!resource_ptr) {
      return zx::error(ZX_ERR_INTERNAL);
    }
    parent_nodes.emplace_back(resource_ptr->owner());
  }

  auto result = Node::CreateCompositeNode(std::string(name), parent_nodes, parent_names_.value(),
                                          parent_properties_, node_manager, dispatcher,
                                          driver_host_name_for_colocation_, primary_index_.value());
  if (result.is_error()) {
    return result.take_error();
  }

  fdf_log::info("Built composite node '{}' for completed composite node spec", name);
  completed_composite_node_.emplace(result.value());
  return zx::ok(result.value());
}

fuchsia_driver_development::wire::CompositeNodeInfo ParentSetCollector::GetCompositeInfo(
    fidl::AnyArena& arena,
    const std::optional<fuchsia_driver_framework::CompositeInfo>& composite_info) const {
  namespace fdd = fuchsia_driver_development;

  auto composite_node_info = fdd::wire::CompositeNodeInfo::Builder(arena);

  if (composite_info.has_value()) {
    composite_node_info.composite(fdd::wire::CompositeInfo::WithComposite(
        arena, fidl::ToWire(arena, composite_info.value())));
  }

  if (!HasCompositeInfo()) {
    fidl::VectorView<fidl::StringView> parent_topological_paths(arena, size());
    composite_node_info.parent_topological_paths(parent_topological_paths);
    composite_node_info.parent_monikers(fidl::VectorView<fidl::StringView>(arena, size()));
    return composite_node_info.Build();
  }

  composite_node_info.parent_topological_paths(GetParentTopologicalPaths(arena));
  composite_node_info.parent_monikers(GetParentMonikers(arena));

  std::optional<NodeWkPtr> composite_node = completed_composite_node();
  if (composite_node) {
    if (auto node_ptr = composite_node->lock(); node_ptr) {
      composite_node_info.topological_path(node_ptr->MakeTopologicalPath());
      composite_node_info.moniker(fidl::StringView(arena, node_ptr->MakeComponentMoniker()));
    }
  }
  return composite_node_info.Build();
}

fidl::VectorView<fidl::StringView> ParentSetCollector::GetParentTopologicalPaths(
    fidl::AnyArena& arena) const {
  fidl::VectorView<fidl::StringView> parent_topological_paths(arena, parents_.size());
  for (uint32_t i = 0; i < parents_.size(); i++) {
    if (parents_[i] == std::nullopt) {
      parent_topological_paths[i] = fidl::StringView();
      continue;
    }

    if (auto resource = parents_[i]->lock(); resource) {
      if (auto node = resource->owner().lock(); node) {
        parent_topological_paths[i] = fidl::StringView(arena, node->MakeTopologicalPath());
      } else {
        parent_topological_paths[i] = fidl::StringView();
      }
    } else {
      parent_topological_paths[i] = fidl::StringView();
    }
  }
  return parent_topological_paths;
}

fidl::VectorView<fidl::StringView> ParentSetCollector::GetParentMonikers(
    fidl::AnyArena& arena) const {
  fidl::VectorView<fidl::StringView> parent_monikers(arena, parents_.size());
  for (uint32_t i = 0; i < parents_.size(); i++) {
    if (parents_[i] == std::nullopt) {
      parent_monikers[i] = fidl::StringView();
      continue;
    }

    if (auto resource = parents_[i]->lock(); resource) {
      if (auto node = resource->owner().lock(); node) {
        parent_monikers[i] = fidl::StringView(arena, node->MakeComponentMoniker());
      } else {
        parent_monikers[i] = fidl::StringView();
      }
    } else {
      parent_monikers[i] = fidl::StringView();
    }
  }
  return parent_monikers;
}

}  // namespace driver_manager

File: src/devices/bin/driver_manager_rust/composite/src/parent_set_collector.rs
// Copyright 2026 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use driver_manager_node::{Node, NodeManager};
use fidl_fuchsia_driver_framework as fdf;
use log::info;
use std::rc::{Rc, Weak};

pub(crate) struct ParentSetCollector {
    composite_name: String,
    parents: Vec<Weak<Node>>,
    parent_names: Vec<String>,
    parent_properties: Vec<fdf::NodePropertyEntry2>,
    primary_index: u32,
    completed_composite_node: Option<Weak<Node>>,
    driver_host_name_for_colocation: String,
}

impl ParentSetCollector {
    pub(crate) fn new(
        composite_name: String,
        parent_names: Vec<String>,
        primary_index: u32,
        driver_host_name_for_colocation: String,
    ) -> Self {
        let num_parents = parent_names.len();
        Self {
            composite_name,
            parents: {
                let mut v = Vec::with_capacity(num_parents);
                for _ in 0..num_parents {
                    v.push(Weak::new());
                }
                v
            },
            parent_names,
            parent_properties: vec![
                fdf::NodePropertyEntry2 {
                    name: "".to_string(),
                    properties: vec![],
                };
                num_parents
            ],
            primary_index,
            completed_composite_node: None,
            driver_host_name_for_colocation,
        }
    }

    pub(crate) fn add_node(
        &mut self,
        index: u32,
        node_properties: Vec<fdf::NodeProperty2>,
        node: Weak<Node>,
    ) -> Result<(), zx::Status> {
        let index = index as usize;
        if index >= self.parents.len() {
            return Err(zx::Status::OUT_OF_RANGE);
        }
        if self.parents[index].upgrade().is_some() {
            return Err(zx::Status::ALREADY_BOUND);
        }
        self.parents[index] = node.clone();
        self.parent_properties[index] = fdf::NodePropertyEntry2 {
            name: self.parent_names[index].clone(),
            properties: node_properties,
        };

        if let Some(node_ptr) = node.upgrade() {
            node_ptr.mark_as_composite_parent();
        }

        Ok(())
    }

    pub(crate) fn release_nodes(&self) {
        for node in &self.parents {
            if let Some(node_ptr) = node.upgrade() {
                node_ptr.unmark_as_composite_parent();
            }
        }
    }

    pub(crate) fn try_to_assemble(
        &mut self,
        node_manager: Box<dyn NodeManager>,
    ) -> Result<Rc<Node>, zx::Status> {
        if let Some(node) = &self.completed_composite_node
            && node.upgrade().is_some()
        {
            return Err(zx::Status::ALREADY_EXISTS);
        }

        if self.parents.iter().any(|node| node.upgrade().is_none()) {
            return Err(zx::Status::SHOULD_WAIT);
        }

        let result = Node::create_composite_node(
            &self.composite_name,
            self.parents.clone(),
            self.parent_names.clone(),
            &self.parent_properties,
            node_manager,
            self.driver_host_name_for_colocation.clone(),
            self.primary_index,
        );

        match result {
            Ok(node) => {
                info!(
                    "Built composite node '{}' for completed composite node spec",
                    self.composite_name
                );
                self.completed_composite_node = Some(Rc::downgrade(&node));
                Ok(node)
            }
            Err(e) => Err(e),
        }
    }

    pub(crate) fn get_parent_topological_paths(&self) -> Vec<Option<String>> {
        self.parents
            .iter()
            .map(|node| node.upgrade().map(|n| n.make_topological_path(false)))
            .collect()
    }

    pub(crate) fn completed_composite_node(&self) -> Option<Weak<Node>> {
        self.completed_composite_node.clone()
    }
}
